In [4]:
import pandas as pd
import cv2
from thefuzz import process, fuzz


In [5]:
import pandas as pd

# Load the datasets from the 'data' folder
# Note: Ensure the filenames match exactly what you have in your folder
tapas_df = pd.read_csv('data/tapas_data.csv') 


# Displaying the structure of the data to identify key columns
print("Tapas Dataset Columns:")
print(tapas_df.columns)

# Preview the first few rows
tapas_df.head()

Tapas Dataset Columns:
Index(['title', 'item_id', 'link', 'cover', 'creators', 'genres', 'views',
       'subscribers', 'likes', 'banner', 'details', 'tags', 'episodes',
       'released'],
      dtype='str')


,title,item_id,link,cover,creators,genres,views,subscribers,likes,banner,details,tags,episodes,released
0,My Gentle Giant,140273,https://tapas.io/series/My-Gentle-Giant/info,https://d30womf5coomej.cloudfront.net/sa/55/a7...,['EmAuthor'],"['BL', 'LGBTQ+', 'Slice of life']","29,109,978 views","241,427 subscribers","2,934,170 likes",https://d30womf5coomej.cloudfront.net/sa/01/c4...,Jun Watanabe is your average outcast. He's a t...,"['#gay', '#soft', '#comedy', '#Angst', '#bl', ...",176,"Sep 28, 2020"
1,DaiMaou,36492,https://tapas.io/series/daimaou/info,https://d30womf5coomej.cloudfront.net/sa/ee/47...,['Amanduur'],"['BL', 'Comedy', 'Fantasy']","24,186,270 views","103,218 subscribers","2,453,503 likes",https://d30womf5coomej.cloudfront.net/sa/e9/5b...,TL;DR: Shitty comedy masquerading as an actual...,"['#gay', '#Fantasy', '#romance', '#comedy', '#...",479,"Jul 08, 2020"
2,Idiots Don't Catch Colds,67447,https://tapas.io/series/Idiots-Dont-Catch-Cold...,https://d30womf5coomej.cloudfront.net/sa/6e/68...,['Aina Palm'],['BL'],"18,534,129 views","132,648 subscribers","1,851,428 likes",https://d30womf5coomej.cloudfront.net/sa/49/92...,There is only one guy Souta absolutely can't s...,"['#romance', '#drama', '#comedy', '#Soccer', '...",233,"Sep 02, 2020"
3,Jamie,110007,https://tapas.io/series/Jamie/info,https://d30womf5coomej.cloudfront.net/sa/a9/05...,"['Bre Indigo', 'Tami']","['LGBTQ+', 'Drama', 'Slice of life']","14,263,432 views","126,167 subscribers","1,452,529 likes",https://d30womf5coomej.cloudfront.net/sa/94/ce...,[ Coming of Age | LGBTQ+ | Young Adult ]\r\n\r...,"['#friendship', '#queer', '#crush', '#lgbt', '...",138,"Jun 04, 2020"
4,FANGS,155459,https://tapas.io/series/fangscomic/info,https://d30womf5coomej.cloudfront.net/sa/18/a9...,['Sarah Andersen'],"['Romance', 'Comedy']","36,237,626 views","160,308 subscribers","1,343,073 likes",6,Vamp is three hundred years old but in all tha...,[],78,"Oct 31, 2019"


In [6]:

webtoon_df = pd.read_csv('data/webtoon_data.csv')
print("\nWebtoons Dataset Columns:")
print(webtoons_df.columns)

webtoon_df.head()


Webtoons Dataset Columns:
Index(['webtoon_id', 'title', 'genre', 'thumbnail', 'summary', 'episodes',
       'Created by', 'view', 'subscribe', 'grade', 'released_date', 'url',
       'cover', 'likes', 'Written by', 'Art by', 'Adapted by',
       'Original work by', 'Assisted by'],
      dtype='str')


,webtoon_id,title,genre,thumbnail,summary,episodes,Created by,view,subscribe,grade,released_date,url,cover,likes,Written by,Art by,Adapted by,Original work by,Assisted by
0,1218,Let's Play,Romance,https://webtoon-phinf.pstatic.net/20210629_103...,"She’s young, single and about to achieve her d...",171,Leeanne M. Krecic (Mongie),606.5M,4.6M,9.59,"Nov 6, 2017",https://www.webtoons.com/en/romance/letsplay/l...,https://webtoon-phinf.pstatic.net/20210629_163...,37.2M,NaN,NaN,NaN,NaN,NaN
1,1436,True Beauty,Romance,https://webtoon-phinf.pstatic.net/20210129_175...,"After binge-watching beauty videos online, a s...",197,Yaongyi,874M,7M,9.53,"Aug 15, 2018",https://www.webtoons.com/en/romance/truebeauty...,https://webtoon-phinf.pstatic.net/20210129_65/...,46.2M,NaN,NaN,NaN,NaN,NaN
2,2135,The Remarried Empress,Fantasy,https://webtoon-phinf.pstatic.net/20200904_29/...,Navier Ellie Trovi was an empress perfect in e...,110,NaN,231.3M,2.5M,9.87,"Sep 5, 2020",https://www.webtoons.com/en/fantasy/the-remarr...,https://webtoon-phinf.pstatic.net/20200904_268...,21.2M,Alphatart,Sumpul,HereLee,NaN,NaN
3,1798,Midnight Poppy Land,Romance,https://webtoon-phinf.pstatic.net/20191119_132...,After making a grisly discovery in the country...,99,Lilydusk,198.8M,2.3M,9.80,"Nov 22, 2019",https://www.webtoons.com/en/romance/midnight-p...,https://webtoon-phinf.pstatic.net/20191119_163...,13.5M,NaN,NaN,NaN,NaN,NaN
4,3416,Reunion,Romance,https://webtoon-phinf.pstatic.net/20220311_196...,"After moving away for a decade, Rhea returns t...",9,stephattyy,7.1M,"629,872",9.77,"Mar 17, 2022",https://www.webtoons.com/en/romance/reunion/li...,https://webtoon-phinf.pstatic.net/20220311_14/...,"570,151",NaN,NaN,NaN,NaN,NaN


In [7]:
print(tapas_df['cover'].iloc[0])

https://d30womf5coomej.cloudfront.net/sa/55/a77bd21b-e4f6-4934-b316-0b9af3cb7bcc_z.jpg


In [8]:
print(webtoon_df['cover'].iloc[0])

https://webtoon-phinf.pstatic.net/20210629_163/1624917409508Qn4m7_JPEG/0LetsPlay_desktop_thumbnail.jpg?type=a210


In [9]:

def create_master_list_with_stats(df_base, df_to_add, threshold=90):
    """
    Unified dataset creator that tracks matching statistics.
    """
    base_titles = df_base['title'].tolist()
    titles_to_check = df_to_add['title'].tolist()
    
    unique_entries = []
    same_titles_found = [] # To track duplicates/matches
    unique_titles_found = [] # To track new unique entries
    
    print(f"Comparing {len(titles_to_check)} titles against base of {len(base_titles)}...")
    
    for title in titles_to_check:
        # Fuzzy Matching logic to align titles 
        match, score = process.extractOne(title, base_titles, scorer=fuzz.token_sort_ratio)
        
        if score >= threshold:
            # This is a 'Same' title (Duplicate across platforms)
            same_titles_found.append({'original': title, 'matched_to': match, 'score': score})
        else:
            # This is a 'Unique' title
            new_row = df_to_add[df_to_add['title'] == title].iloc[0]
            unique_entries.append(new_row)
            unique_titles_found.append(title)
            
    # Combine to create the unified dataset 
    if unique_entries:
        additional_df = pd.DataFrame(unique_entries)
        master_df = pd.concat([df_base, additional_df], ignore_index=True)
    else:
        master_df = df_base.copy()

    # Final Stats Output
    print("-" * 30)
    print(f"MATCHING RESULTS:")
    print(f"Total 'Same' titles found (Duplicates): {len(same_titles_found)}")
    print(f"Total 'Unique' titles added: {len(unique_titles_found)}")
    print(f"Final Dataset Size: {len(master_df)}")
    print("-" * 30)
    
    # Returning the lists so you can inspect them if you want
    return master_df, same_titles_found

# Run the function
master_list, duplicates_log = create_master_list_with_stats(webtoons_df, tapas_df)

# If you want to see a few examples of what it thought was the 'same':
print("Sample of matches found:")
print(pd.DataFrame(duplicates_log).head(10))

Comparing 737 titles against base of 734...
------------------------------
MATCHING RESULTS:
Total 'Same' titles found (Duplicates): 2
Total 'Unique' titles added: 735
Final Dataset Size: 1469
------------------------------
Sample of matches found:
    original matched_to  score
0  Nevermore  Nevermore    100
1   Flawless   Flawless    100


In [10]:
import requests
import time
import pandas as pd

def fetch_manhwa_metadata(query_title):
    """
    Retrieves current Manhwa-specific metadata from Jikan API.
    """
    # Endpoint for manga/manhwa search
    url = "https://api.jikan.moe/v4/manga"
    
    # Parameters: Search title, filter by manhwa, and limit results
    params = {
        'q': query_title,
        'type': 'manhwa', # Strictly filter for Korean comics
        'limit': 1
    }
    
    try:
        response = requests.get(url, params=params)
        
        if response.status_code == 200:
            data = response.json().get('data', [])
            if data:
                item = data[0]
                return {
                    'title': item.get('title'),
                    'rank': item.get('rank'),           # Global Ranking
                    'popularity': item.get('popularity'), # Popularity Metric
                    'score': item.get('score'),         # User Score
                    'cover_url': item['images']['jpg']['large_image_url'], # High Res Cover
                    'status': item.get('status')
                }
        elif response.status_code == 429:
            print("Rate limit hit! Sleeping for 5 seconds...")
            time.sleep(5)
            
    except Exception as e:
        print(f"Request error for {query_title}: {e}")
    
    return None

# --- TESTING THE ROADMAP ---
# Test with a known Manhwa from your list
test_result = fetch_manhwa_metadata("Solo Leveling")
print(test_result)

{'title': 'Solo Leveling', 'rank': 122, 'popularity': 5, 'score': 8.58, 'cover_url': 'https://myanimelist.net/images/manga/3/222295l.jpg', 'status': 'Finished'}


In [12]:
import os
import json
import time
import requests
import pandas as pd
from thefuzz import fuzz

# Constants for the enrichment pipeline [cite: 9]
CACHE_DIR = 'data/json_cache'
os.makedirs(CACHE_DIR, exist_ok=True)

def get_sanitized_filename(title):
    """Creates a safe filename for the JSON cache."""
    return "".join(x for x in title if x.isalnum()).lower() + ".json"

def smart_enrichment_loop(master_df, limit=None):
    """
    Efficiently fetches Manhwa metadata using a cache-first approach.
    """
    base_url = "https://api.jikan.moe/v4/manga"
    headers = {'User-Agent': 'ManhwaProject/1.0'}
    
    # Optional limit for testing/sampling 
    subset = master_df.head(limit) if limit else master_df

    for index, row in subset.iterrows():
        original_title = row['title']
        filename = get_sanitized_filename(original_title)
        file_path = os.path.join(CACHE_DIR, filename)

        # 1. Check if data already exists (Efficiency Step)
        if os.path.exists(file_path):
            continue # Skip to next title

        print(f"[{index}] Fetching from API: {original_title}")
        
        try:
            # 2. API Request with Manhwa filter [cite: 6]
            params = {'q': original_title, 'type': 'manhwa', 'limit': 3}
            response = requests.get(base_url, params=params, headers=headers)
            
            if response.status_code == 200:
                results = response.json().get('data', [])
                
                # 3. Fuzzy Matching logic to ensure data integrity [cite: 10, 17]
                best_match = None
                highest_similarity = 0
                
                for item in results:
                    similarity = fuzz.token_sort_ratio(original_title, item['title'])
                    if similarity > highest_similarity:
                        highest_similarity = similarity
                        best_match = item
                
                # 4. Save only if match is confident (>80%)
                if best_match and highest_similarity > 80:
                    metadata = {
                        'original_title': original_title,
                        'mal_id': best_match['mal_id'],
                        'title_api': best_match['title'],
                        'rank': best_match.get('rank'),
                        'popularity': best_match.get('popularity'),
                        'score': best_match.get('score'),
                        'cover_url': best_match['images']['jpg']['large_image_url'],
                        'genres': [g['name'] for g in best_match.get('genres', [])],
                        'match_confidence': highest_similarity
                    }
                    
                    with open(file_path, 'w', encoding='utf-8') as f:
                        json.dump(metadata, f, indent=4)
                
                # 5. Respect Jikan Rate Limits (3 requests/sec)
                time.sleep(1.2) 
                
            elif response.status_code == 429:
                print("Rate limit hit. Cooling down for 30 seconds...")
                time.sleep(30)
                
        except Exception as e:
            print(f"Error processing {original_title}: {e}")
            continue

# Run the smart loop on your master_list 
smart_enrichment_loop(master_list)

[0] Fetching from API: Let's Play
[3] Fetching from API: Midnight Poppy Land
[7] Fetching from API: I Love Yoo
[8] Fetching from API: Age Matters
[10] Fetching from API: Fictional Skin
[12] Fetching from API: Acception
[13] Fetching from API: Edith
[17] Fetching from API: Purple Hyacinth
[23] Fetching from API: Boyfriend of the Dead
[24] Fetching from API: Archie Comics:Big Ethel Energy
[28] Fetching from API: Love Me Knot
[40] Fetching from API: Mistake
[42] Fetching from API: Lost in Translation
[48] Fetching from API: Seven Years Later
[53] Fetching from API: Mom, I'm Sorry
[56] Fetching from API: Watermelon
[57] Fetching from API: Raven Saga
[58] Fetching from API: Before We Knew It
[59] Fetching from API: Señorita Cometa
[61] Fetching from API: The Croaking
[62] Fetching from API: Blue
[64] Fetching from API: Madame Outlaw
[69] Fetching from API: Who's Mr. President?
[70] Fetching from API: Woven
[71] Fetching from API: There's Love Hidden in Lies
[72] Fetching from API: Unwritten

In [13]:
# --- EXECUTION CELL ---

# 1. Run the smart loop for the first 20 titles only as a 'Smoke Test'
# This implements the multi-source enrichment pipeline 
print("Starting API Enrichment...")
smart_enrichment_loop(master_list, limit=20)

# 2. After the loop finishes, consolidate the results
enriched_df = finalize_unified_dataset()

# 3. Check if we successfully captured the 'Numerical Features' [cite: 15]
if not enriched_df.empty:
    print(f"Successfully enriched {len(enriched_df)} titles.")
    display(enriched_df[['original_title', 'rank', 'score', 'cover_url']].head())
else:
    print("No data found yet. Check your internet connection or API response.")

Starting API Enrichment...
[0] Fetching from API: Let's Play
[3] Fetching from API: Midnight Poppy Land
[7] Fetching from API: I Love Yoo
[8] Fetching from API: Age Matters
[10] Fetching from API: Fictional Skin
[12] Fetching from API: Acception
[13] Fetching from API: Edith
[17] Fetching from API: Purple Hyacinth


NameError: name 'finalize_unified_dataset' is not defined

In [ ]:
import pandas as pd

def load_and_clean_kaggle_data(file_path):
    """
    Loads CSV data while handling broken lines caused by newlines and commas.
    """
    try:
        # on_bad_lines='warn' tells us which lines are broken
        # engine='python' is more flexible for messy files
        df = pd.read_csv(
            file_path, 
            quotechar='"', 
            skipinitialspace=True,
            on_bad_lines='warn', 
            engine='python'
        )
        
        # Clean up the text columns: Remove actual \n characters that 
        # made it into the dataframe cells
        text_columns = df.select_dtypes(include=['object']).columns
        for col in text_columns:
            df[col] = df[col].astype(str).str.replace(r'\n', ' ', regex=True).str.strip()
            
        print(f"Successfully loaded {len(df)} rows from {file_path}")
        return df
    except Exception as e:
        print(f"Error loading {file_path}: {e}")
        return None

# Apply the fix to your specific files
tapas_df = load_and_clean_kaggle_data('data/tapas_data.csv')
webtoons_df = load_and_clean_kaggle_data('data/webtoon_data.csv')

# Verify the column separation
print(tapas_df.columns)
tapas_df.head()

Successfully loaded 737 rows from data/tapas_data.csv
Successfully loaded 734 rows from data/webtoon_data.csv
Index(['title', 'item_id', 'link', 'cover', 'creators', 'genres', 'views',
       'subscribers', 'likes', 'banner', 'details', 'tags', 'episodes',
       'released'],
      dtype='str')


/var/folders/6n/b4blqqt925d3qy7nk2b4z5200000gn/T/ipykernel_49046/3016862315.py:20: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  text_columns = df.select_dtypes(include=['object']).columns
/var/folders/6n/b4blqqt925d3qy7nk2b4z5200000gn/T/ipykernel_49046/3016862315.py:20: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https

,title,item_id,link,cover,creators,genres,views,subscribers,likes,banner,details,tags,episodes,released
0,My Gentle Giant,140273,https://tapas.io/series/My-Gentle-Giant/info,https://d30womf5coomej.cloudfront.net/sa/55/a7...,['EmAuthor'],"['BL', 'LGBTQ+', 'Slice of life']","29,109,978 views","241,427 subscribers","2,934,170 likes",https://d30womf5coomej.cloudfront.net/sa/01/c4...,Jun Watanabe is your average outcast. He's a t...,"['#gay', '#soft', '#comedy', '#Angst', '#bl', ...",176,"Sep 28, 2020"
1,DaiMaou,36492,https://tapas.io/series/daimaou/info,https://d30womf5coomej.cloudfront.net/sa/ee/47...,['Amanduur'],"['BL', 'Comedy', 'Fantasy']","24,186,270 views","103,218 subscribers","2,453,503 likes",https://d30womf5coomej.cloudfront.net/sa/e9/5b...,TL;DR: Shitty comedy masquerading as an actual...,"['#gay', '#Fantasy', '#romance', '#comedy', '#...",479,"Jul 08, 2020"
2,Idiots Don't Catch Colds,67447,https://tapas.io/series/Idiots-Dont-Catch-Cold...,https://d30womf5coomej.cloudfront.net/sa/6e/68...,['Aina Palm'],['BL'],"18,534,129 views","132,648 subscribers","1,851,428 likes",https://d30womf5coomej.cloudfront.net/sa/49/92...,There is only one guy Souta absolutely can't s...,"['#romance', '#drama', '#comedy', '#Soccer', '...",233,"Sep 02, 2020"
3,Jamie,110007,https://tapas.io/series/Jamie/info,https://d30womf5coomej.cloudfront.net/sa/a9/05...,"['Bre Indigo', 'Tami']","['LGBTQ+', 'Drama', 'Slice of life']","14,263,432 views","126,167 subscribers","1,452,529 likes",https://d30womf5coomej.cloudfront.net/sa/94/ce...,[ Coming of Age | LGBTQ+ | Young Adult ]\r \r ...,"['#friendship', '#queer', '#crush', '#lgbt', '...",138,"Jun 04, 2020"
4,FANGS,155459,https://tapas.io/series/fangscomic/info,https://d30womf5coomej.cloudfront.net/sa/18/a9...,['Sarah Andersen'],"['Romance', 'Comedy']","36,237,626 views","160,308 subscribers","1,343,073 likes",6,Vamp is three hundred years old but in all tha...,[],78,"Oct 31, 2019"


In [14]:
import pandas as pd

def create_integrated_df(master_list, tapas_df, webtoons_df, jikan_cache_df):
    """
    Creates a unified dataset with source tracking and 4-year comparison flags.
    """
    # 1. Normalize titles for comparison
    tapas_set = set(tapas_df['title'].str.lower().unique())
    webtoon_set = set(webtoons_df['title'].str.lower().unique())
    jikan_set = set(jikan_cache_df['original_title'].str.lower().unique())

    # 2. Function to determine source origin
    def get_source(title):
        t = title.lower()
        sources = []
        if t in tapas_set: sources.append("Tapas")
        if t in webtoon_set: sources.append("Webtoon Platform")
        if t in jikan_set: sources.append("Jikan API")
        return " & ".join(sources) if sources else "Unknown"

    # 3. Build the Integrated DataFrame
    integrated_df = master_list.copy()
    integrated_df['data_origin'] = integrated_df['title'].apply(get_source)
    
    # 4. Merge Jikan Metrics (Current popularity and rank)
    integrated_df = pd.merge(
        integrated_df, 
        jikan_cache_df[['original_title', 'rank', 'score', 'cover_url']], 
        left_on='title', 
        right_on='original_title', 
        how='left'
    )
    
    return integrated_df

# Execute
final_df = create_integrated_df(master_list, tapas_df, webtoons_df, jikan_metadata_df)

NameError: name 'jikan_metadata_df' is not defined

master_list = combination of two kaggle datasets
json_cache is a directory


This code find the number of manhwa's included in kaggle datasets but couldn't find with API

In [20]:
import os
import pandas as pd

def audit_json_cache(master_list, cache_dir='data/json_cache'):
    """
    Compares the master list against the JSON cache to identify missing data.
    """
    # 1. Get the list of all sanitized filenames currently in the cache
    cached_files = {f.replace('.json', '') for f in os.listdir(cache_dir) if f.endswith('.json')}
    
    # 2. Extract all titles from your master_list
    all_titles = master_list['title'].tolist()
    
    matched_titles = []
    missing_titles = []

    for title in all_titles:
        # We use the same sanitization logic used during the fetching process
        sanitized_name = "".join(x for x in title if x.isalnum()).lower()
        
        if sanitized_name in cached_files:
            matched_titles.append(title)
        else:
            missing_titles.append(title)

    # 3. Calculate statistics
    total = len(all_titles)
    found = len(matched_titles)
    missing = len(missing_titles)
    success_rate = (found / total) * 100

    print("--- DATASET AUDIT REPORT ---")
    print(f"Total Kaggle Titles:   {total}")
    print(f"Successfully Matched:  {found}")
    print(f"Missing from Cache:    {missing}")
    print(f"Coverage Success:      {success_rate:.2f}%")
    print("-" * 28)

    # Save the missing titles to a CSV for later inspection (e.g., the 'Let's Play' cases)
    missing_df = pd.DataFrame(missing_titles, columns=['missing_title'])
    missing_df.to_csv('data/missing_from_jikan.csv', index=False)
    print("List of missing titles saved to 'data/missing_from_jikan.csv'")
    
    return missing_titles

# Execute the audit
missing_manhwa = audit_json_cache(master_list)

--- DATASET AUDIT REPORT ---
Total Kaggle Titles:   1469
Successfully Matched:  723
Missing from Cache:    746
Coverage Success:      49.22%
----------------------------
List of missing titles saved to 'data/missing_from_jikan.csv'


Step 1: Create the Ultimate DataFrame
First, we merge the Kaggle data with the Jikan metadata to have all URLs in one row.

In [22]:

def create_ultimate_research_df(master_list, cache_dir='data/json_cache'):
    """
    Integrates Kaggle URLs and Jikan Metadata into a single research-ready DataFrame.
    """
    # Load Jikan metadata from JSON files
    json_list = []
    for filename in os.listdir(cache_dir):
        if filename.endswith('.json'):
            with open(os.path.join(cache_dir, filename), 'r', encoding='utf-8') as f:
                json_list.append(json.load(f))
    
    jikan_df = pd.DataFrame(json_list)
    
    # Inner Join: Only keep titles that exist in both Kaggle and Jikan Cache
    # We rename columns to be explicit about the timeline
    ultimate_df = pd.merge(
        master_list[['title', 'cover']], # 'cover' is the Kaggle 2022 URL
        jikan_df[['original_title', 'cover_url', 'score', 'rank', 'popularity']], 
        left_on='title', 
        right_on='original_title', 
        how='inner' #Inner Join only keeps rows where the title exists in 
        #both the Kaggle list and the Jikan Cache.
    )
    
    # Rename for clarity
    ultimate_df = ultimate_df.rename(columns={
        'cover': 'url_2022',
        'cover_url': 'url_2026'
    })
    
    # Drop redundant column
    if 'original_title' in ultimate_df.columns:
        ultimate_df.drop(columns=['original_title'], inplace=True)
        
    print(f"Ultimate DataFrame created with {len(ultimate_df)} titles.")
    return ultimate_df

# Execute
ultimate_df = create_ultimate_research_df(master_list)

Ultimate DataFrame created with 723 titles.


ultimate_df contains only 723 manhwa/webtoon which both in MAL and 2 kaggle datasets.

Step 2: The Dual-Version Image Downloader
This function implements your logic: it tries to download the 2022 version (using headers) and the 2026 version. It tracks which ones were successful.

In [23]:
import requests
import time

def download_dual_covers(df):
    """
    Downloads cover images for both 2022 (Kaggle) and 2026 (Jikan) versions.
    Tracks success status for data quality reporting.
    """
    base_dir = 'data/covers'
    dir_2022 = os.path.join(base_dir, 'v2022')
    dir_2026 = os.path.join(base_dir, 'v2026')
    
    os.makedirs(dir_2022, exist_ok=True)
    os.makedirs(dir_2026, exist_ok=True)
    
    # Essential headers to bypass server-side blocks
    headers = {
        'Referer': 'https://www.webtoons.com/',
        'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36'
    }

    # Tracking lists to update the DataFrame later
    status_2022 = []
    status_2026 = []

    for index, row in df.iterrows():
        title = row['title']
        # Create a safe filename
        safe_name = "".join(x for x in title if x.isalnum()).lower() + ".jpg"
        
        path_2022 = os.path.join(dir_2022, safe_name)
        path_2026 = os.path.join(dir_2026, safe_name)

        # 1. Process 2022 Image (Kaggle)
        success_22 = False
        if os.path.exists(path_2022):
            success_22 = True
        elif pd.notna(row['url_2022']):
            try:
                resp = requests.get(row['url_2022'], headers=headers, timeout=10)
                if resp.status_code == 200:
                    with open(path_2022, 'wb') as f:
                        f.write(resp.content)
                    success_22 = True
            except:
                pass
        
        # 2. Process 2026 Image (Jikan)
        success_26 = False
        if os.path.exists(path_2026):
            success_26 = True
        elif pd.notna(row['url_2026']):
            try:
                resp = requests.get(row['url_2026'], headers=headers, timeout=10)
                if resp.status_code == 200:
                    with open(path_2026, 'wb') as f:
                        f.write(resp.content)
                    success_26 = True
            except:
                pass

        status_2022.append(success_22)
        status_2026.append(success_26)
        
        if index % 50 == 0:
            print(f"Progress: {index}/{len(df)} titles processed.")
        
        # Respect servers to avoid getting banned
        time.sleep(0.2)

    # Add status flags to the DataFrame
    df['has_v2022'] = status_2022
    df['has_v2026'] = status_2026
    
    return df

# Start the download process
ultimate_df = download_dual_covers(ultimate_df)

Progress: 0/723 titles processed.
Progress: 50/723 titles processed.
Progress: 100/723 titles processed.
Progress: 150/723 titles processed.
Progress: 200/723 titles processed.
Progress: 250/723 titles processed.
Progress: 300/723 titles processed.
Progress: 350/723 titles processed.
Progress: 400/723 titles processed.
Progress: 450/723 titles processed.
Progress: 500/723 titles processed.
Progress: 550/723 titles processed.
Progress: 600/723 titles processed.
Progress: 650/723 titles processed.
Progress: 700/723 titles processed.


Run this code after the downloader finishes. It will tell you exactly what is missing and why.

In [24]:
# 1. Filter the rows where the Kaggle (2022) download failed
broken_2022_df = ultimate_df[ultimate_df['has_v2022'] == False]

# 2. Filter the rows where the Jikan (2026) download failed
broken_2026_df = ultimate_df[ultimate_df['has_v2026'] == False]

# 3. Print the results
print("--- BROKEN URL REPORT ---")
print(f"Total Titles Checked: {len(ultimate_df)}")
print(f"Broken Kaggle (2022) URLs: {len(broken_2022_df)}")
print(f"Broken Jikan (2026) URLs:  {len(broken_2026_df)}")
print("-" * 25)

# 4. Display the list of broken Kaggle titles
if not broken_2022_df.empty:
    print("\nSample of Titles with Broken 2022 Links:")
    print(broken_2022_df[['title', 'url_2022']].head(10))
    
    # Optional: Save the broken list to a CSV for your report
    broken_2022_df[['title', 'url_2022']].to_csv('data/broken_links_2022.csv', index=False)

--- BROKEN URL REPORT ---
Total Titles Checked: 723
Broken Kaggle (2022) URLs: 1
Broken Jikan (2026) URLs:  0
-------------------------

Sample of Titles with Broken 2022 Links:
                              title  \
518  The Legendary Spearman Returns   

                                              url_2022  
518  https://d30womf5coomej.cloudfront.net/sa/ea/50...  


In [25]:
# Group A: Perfect matches (Before & After comparison possible)
comparable_df = ultimate_df[(ultimate_df['has_v2022'] == True) & (ultimate_df['has_v2026'] == True)]

# Group B: Current only (Only 2026 analysis possible)
current_only_df = ultimate_df[(ultimate_df['has_v2022'] == False) & (ultimate_df['has_v2026'] == True)]

print(f"Titles available for Before/After Analysis: {len(comparable_df)}")
print(f"Titles available for Current Popularity Analysis: {len(ultimate_df[ultimate_df['has_v2026'] == True])}")

Titles available for Before/After Analysis: 722
Titles available for Current Popularity Analysis: 723


comparable_df = manhwa's which urls coming from jikan and kaggle datasets works.

In [26]:
# Create the final clean dataset for OpenCV and Statistics
# Keep only titles where both versions were successfully downloaded
final_analysis_df = ultimate_df[(ultimate_df['has_v2022'] == True) & (ultimate_df['has_v2026'] == True)].copy()

print(f"Final sample size for research: {len(final_analysis_df)}")
final_analysis_df.to_csv('data/final_manhwa_research_data.csv', index=False)

Final sample size for research: 722
